feat: add Linux support

isaiahdahl opened the pull request:
## Summary

This PR adds complete Linux platform support to CodeLayer (the HumanLayer WUI), enabling the application to build, run, and be distributed on Linux systems alongside the existing macOS support.

## Problem

CodeLayer was previously macOS-only:
- The keyboard shortcut used `cmd+shift+h` which doesn't work on Linux
- Build targets were hardcoded for `darwin-arm64`
- No CI/CD pipeline existed for Linux releases
- Documentation only covered macOS development and installation

## Solution

### 1. Cross-Platform Keyboard Shortcut
Changed the quick launcher global shortcut from macOS-specific to cross-platform:
```rust
// Before
shortcut.on_shortcut("cmd+shift+h", ...)

// After  
shortcut.on_shortcut("CommandOrControl+Shift+h", ...)
```
This uses Tauri's `CommandOrControl` modifier which maps to `Cmd` on macOS and `Ctrl` on Linux/Windows.

### 2. Linux Build Targets
Added three new Makefile targets:

| Target | Description |
|--------|-------------|
| `humanlayer-binary-linux-x64` | Builds humanlayer CLI for Linux x86_64 |
| `codelayer-bundle-linux` | Builds daemon + CLI for Linux bundling |
| `codelayer-nightly-bundle-linux` | Complete nightly build with AppImage/deb packages |

### 3. Linux Release Workflow
Created `.github/workflows/release-linux.yml` that:
- Triggers on version tags (`v*.*.*`) or manual dispatch
- Builds `hld-linux-x64` daemon binary
- Builds `humanlayer-linux-x64` CLI via Bun
- Creates AppImage and .deb packages via Tauri
- Uploads artifacts to GitHub Releases

### 4. Documentation Updates
- **CONTRIBUTING.md**: Added Linux development prerequisites with package commands for Debian, Arch, and Fedora
- **humanlayer-wui/README.md**: Updated "Building for Production" to include Linux commands
- **humanlayer-wui/RELEASE.md**: Added comprehensive Linux release workflow documentation

## Files Changed

| File | Changes |
|------|---------|
| `humanlayer-wui/src-tauri/src/lib.rs` | Cross-platform shortcut modifier |
| `Makefile` | +60 lines for Linux build targets |
| `.github/workflows/release-linux.yml` | **New** - 358 lines |
| `CONTRIBUTING.md` | +15 lines for Linux prerequisites |
| `humanlayer-wui/README.md` | +9 lines for Linux build instructions |
| `humanlayer-wui/RELEASE.md` | +58 lines for Linux release docs |

## Linux Packages

When merged, each release will automatically include:
- `CodeLayer-linux-x64.AppImage` - Universal Linux package (works on any distro)
- `codelayer-linux-x64.deb` - Debian/Ubuntu package

## How to Verify

- [x] Rust code compiles on Linux (`cargo check`)
- [x] Daemon builds on Linux (`make daemon-dev-build`)
- [x] Tauri dev build succeeds on Linux
- [x] All WUI checks pass (`make check-wui`)
- [x] All WUI tests pass (`make test-wui`)
- [ ] Full AppImage creation (requires CI run after merge)
- [ ] Full deb package creation (requires CI run after merge)

## Breaking Changes

None. All changes are additive and do not affect the existing macOS build process.

## Notes

- The Rust code already had proper `#[cfg(target_os = "macos")]` guards for macOS-specific features (window background color, appearance)
- Linux-specific Tauri dependencies are handled by the existing `hack/install_platform_deps.sh` script which supports Debian, Arch, Fedora, and openSUSE
- The Linux workflow runs in parallel with the macOS workflow, both attaching artifacts to the same GitHub Release

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Adds Linux support to CodeLayer with cross-platform shortcuts, Linux build targets, CI/CD, and updated documentation.
> 
>   - **Behavior**:
>     - Changed keyboard shortcut from `cmd+shift+h` to `CommandOrControl+Shift+h` in `lib.rs` for cross-platform compatibility.
>     - Added Linux build targets in `Makefile` for `humanlayer-binary-linux-x64`, `codelayer-bundle-linux`, and `codelayer-nightly-bundle-linux`.
>     - Created `.github/workflows/release-linux.yml` for Linux CI/CD, building and releasing AppImage and .deb packages.
>   - **Documentation**:
>     - Updated `CONTRIBUTING.md` with Linux development prerequisites.
>     - Updated `README.md` and `RELEASE.md` to include Linux build and release instructions.
>   - **Misc**:
>     - Minor logging improvements in `lib.rs`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=humanlayer%2Fhumanlayer&utm_source=github&utm_medium=referral)<sup> for f1b2d243aadc4f88c703f178645b62d61fa18989. You can [customize](https://app.ellipsis.dev/humanlayer/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

isaiahdahl committed:
feat: add full Linux support for CodeLayer

Add cross-platform support enabling CodeLayer (WUI) to build and run on Linux.

Changes:
- Use CommandOrControl+Shift+h shortcut for cross-platform compatibility
- Add Linux x64 build targets in Makefile (codelayer-bundle-linux, etc.)
- Add release-linux.yml GitHub Actions workflow for AppImage/deb builds
- Update documentation with Linux development prerequisites and install instructions

The Linux release workflow creates AppImage and .deb packages that are
attached to GitHub Releases alongside the existing macOS DMG.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/humanlayer/humanlayer?pullRequest=913) <br/>All committers have signed the CLA.

isaiahdahl committed:
fix: sanitize branch ID and ensure directory creation for daemon

Branch names with slashes (e.g., 'feat/branch') were causing file copy
failures because the application attempted to create files in non-existent
subdirectories under ~/.humanlayer.

Changes:
- Sanitize branch ID by replacing '/' with '-' in get_branch_id
- Add defensive fs::create_dir_all before copying the dev database

isaiahdahl committed:
feat(linux): improve production bundling and runtime robustness

- Update Makefile bundling targets to use -ldflags, ensuring the daemon
  correctly identifies the production CLI command as 'humanlayer'.
- Fix Makefile to reliably copy fresh binaries into Tauri resources
  during the bundling process.
- Modify daemon runtime to automatically include its resource directory
  in the PATH, allowing seamless discovery of the bundled CLI.
- Sanitize branch identifiers to prevent directory creation failures
  when branch names contain slashes.
- Migrate global shortcuts to use 'CommandOrControl' for cross-platform
  compatibility.
- Fix various clippy lints (uninlined format arguments) in WUI backend.

digi4care commented:
Tested PR #913 on Linux (EndeavourOS / Arch-based).

What I ran (in a clean worktree):
- `make setup`
- `make daemon-dev-build`
- `make check-wui` (format/lint/tsc + cargo check + clippy)
- `make test-wui`
- `make wui-dev` (started, long first-time Rust compile)

Results:
- `make setup` FAILS in a fresh shell because `mockgen` is installed but not on `PATH`.
  - I verified `mockgen` exists at `$(go env GOPATH)/bin/mockgen` (v0.6.0).
  - Workaround: `export PATH="$(go env GOPATH)/bin:$PATH"` then `make setup` passes.
  - This is the same class of issue as earlier Linux bootstrap fixes (PATH for Go-installed tools).
- `make setup` prints `Unsupported distribution: endeavouros`.
  - On EndeavourOS we want it treated as Arch (docs or mapping).
- With `PATH` workaround in place:
  - `make setup` ✅
  - `make daemon-dev-build` ✅
  - `make check-wui` ✅ (cargo check + clippy -D warnings OK)
  - `make test-wui` ✅
- `make wui-dev` started compiling dependencies; I didn’t reach the running window within my time limit (no functional failure observed, just long initial compile).

Suggestion for improving Linux dev UX:
- Add `$(go env GOPATH)/bin` (or `GOBIN`) to `PATH` inside setup scripts when installing tools like `mockgen`.
- Map `endeavouros -> arch` (or document it) so the distro check doesn’t warn unnecessarily.
- Consider documenting Wayland rendering workaround if relevant: `WEBKIT_DISABLE_DMABUF_RENDERER=1` and optionally `GDK_BACKEND=x11` (I’ve needed this on Wayland with WebKitGTK).

digi4care commented:
Tested Linux build targets on EndeavourOS (x86_64).

Env:
- bun 1.3.5
- go go1.25.5 linux/amd64
- rustc 1.92.0 (LLVM 21.1.6)

✅ make humanlayer-binary-linux-x64
- produced: hlyr/humanlayer-linux-x64 (~104.7MB)
- ./hlyr/humanlayer-linux-x64 --help works

✅ make codelayer-bundle-linux
- produced: hld/hld-linux-x64 (~28.8MB)
- produced: hlyr/humanlayer-linux-x64
- copied to Tauri resources:
  - humanlayer-wui/src-tauri/bin/hld
  - humanlayer-wui/src-tauri/bin/humanlayer

⚠️ make codelayer-nightly-bundle-linux (AppImage/DEB)
- Got past bun/tsc/vite build and Rust release build.
- Still fails at bundling step with:
  - "failed to bundle project: failed to run linuxdeploy"
- Also see Vite/tsc warning during build:
  - Cannot find base config file "@codelayer/typescript-config/base.json" (../tsconfig.json)

Notes:
- I initially hit a rustc SIGSEGV while compiling gtk/pango-sys in release; rerun with:
  APPIMAGE_EXTRACT_AND_RUN=1 RUST_MIN_STACK=16777216 CARGO_BUILD_JOBS=1 CARGO_PROFILE_RELEASE_OPT_LEVEL=1
  avoided the rustc crash, but linuxdeploy bundling still fails.

digi4care commented:
Tested Linux build targets on EndeavourOS (x86_64).

Env:
- bun 1.3.5
- go go1.25.5 linux/amd64
- rustc 1.92.0 (LLVM 21.1.6)

✅ make humanlayer-binary-linux-x64
- produced: hlyr/humanlayer-linux-x64 (~104.7MB)
- ./hlyr/humanlayer-linux-x64 --help works

✅ make codelayer-bundle-linux
- produced: hld/hld-linux-x64 (~28.8MB)
- produced: hlyr/humanlayer-linux-x64
- copied to Tauri resources:
  - humanlayer-wui/src-tauri/bin/hld
  - humanlayer-wui/src-tauri/bin/humanlayer

✅ AppImage bundling (linuxdeploy)
- produced:
  - humanlayer-wui/src-tauri/target/release/bundle/appimage/CodeLayer-Nightly-x86_64.AppImage
  - size: 245,589,184 bytes
  - sha256: d144276827c87a9e292a66371410266b1d6800121b90c61f0b3a2aa76b7d19bf
- Note: I did not get a `.deb` output directory (`.../bundle/deb/` was not created).

Bundling details/workarounds used:
- `NO_STRIP=1` (otherwise linuxdeploy/strip chokes on `.relr.dyn` on Arch)
- `APPIMAGE_EXTRACT_AND_RUN=1`
- `ARCH=x86_64` (appimagetool otherwise complains about multiple arch)
- bun sidecar binary (`humanlayer`) gets corrupted when linuxdeploy sets RUNPATH/RPATH; workaround:
  - extract linuxdeploy AppImage and add a small wrapper around embedded `patchelf` to skip `--set-rpath` for `.../CodeLayer-Nightly/bin/humanlayer`
- gtk plugin reruns failed on existing symlinks; workaround:
  - patch `~/.cache/tauri/linuxdeploy-plugin-gtk.sh` to use `ln -sf ...` (instead of `ln -s ...`)

Logs:
- `/tmp/linuxdeploy-codelayer-final.log`

Known warning during build:
- Vite/tsc warning:
  - Cannot find base config file "@codelayer/typescript-config/base.json" (../tsconfig.json)

Runtime note (Arch / KDE6 / Wayland / Nvidia):
- The AppImage only starts reliably with WebKitGTK software rendering workarounds, e.g.:
  - `WEBKIT_DISABLE_COMPOSITING_MODE=1 ./CodeLayer-Nightly-x86_64.AppImage`
  - or (full software) `LIBGL_ALWAYS_SOFTWARE=1 ./CodeLayer-Nightly-x86_64.AppImage`

Notes:
- I initially hit a rustc SIGSEGV while compiling gtk/pango-sys in release; rerun with:
  APPIMAGE_EXTRACT_AND_RUN=1 RUST_MIN_STACK=16777216 CARGO_BUILD_JOBS=1 CARGO_PROFILE_RELEASE_OPT_LEVEL=1
  avoided the rustc crash.

isaiahdahl committed:
fix(linux): add GOPATH/bin to PATH and support Arch derivatives

- Add GOPATH/bin to PATH in setup_repo.sh to fix mockgen not found issue
- Expand Arch Linux detection to include EndeavourOS, Garuda, Artix, Arcolinux, and Cachyos
- Fixes 'make setup' failure on fresh Linux shells where Go-installed binaries aren't in PATH

isaiahdahl committed:
docs(linux): add Wayland troubleshooting and build workarounds

- Add Wayland/WebKitGTK troubleshooting section to CONTRIBUTING.md for developers
- Document WebKitGTK rendering workarounds (DMABUF, X11, software rendering)
- Add Linux local build workarounds section to RELEASE.md
- Document known linuxdeploy/patchelf issues with bun binaries

isaiahdahl committed:
fix(linux): add AppImage build workarounds and fix tsconfig resolution

- Add NO_STRIP=1 and APPIMAGE_EXTRACT_AND_RUN=1 env vars to Linux builds
- Prevents strip failures on binaries with .relr.dyn sections (Arch-based distros)
- Allows AppImage tools to run without FUSE in containerized builds
- Fix tsconfig.json resolution warning by inlining base config instead of extending package

isaiahdahl commented:
@digi4care thanks for the feedback and thorough testing
